### PR TITLE
PARQUET-1300: [C++] Implement encrypted Parquet read and write support

### PR DESCRIPTION
**Pull request by thamht4190**
I open a fresh pull request from the pull #2555 since pull #2555 is too big now. Please let your comments here. Thanks!

### CODE REVIEW

File: cpp/src/parquet/file_reader.cc
Comment (majetideepak) [inline cpp/src/parquet/file_reader.cc]: Can we remove these `ifdef` and use conditionals based on the presence of crypto variables? In this case we can check based off `col->crypto_metadata()`?

**Comment by revit1976** [comment]
@wesm , @majetideepak  fyi,  all CI issues are now resolved.

File: cpp/src/parquet/file_reader.cc
Comment (ggershinsky) [reply cpp/src/parquet/file_reader.cc]: Yep, looks like col->crypto_metadata() check does the job here. We'll also handle all other similar ifdefs.

Comment (thamht4190) [reply cpp/src/parquet/file_reader.cc]: I have removed some but not completely. I'm not sure if it matches your expectation. Can you please take a look again?

**Comment by wesm** [comment]
I won't be able to do a thorough review but would like to get this merged soon. Does the PR into parquet-testing need to be merged before this is?

**Comment by ggershinsky** [comment]
> Does the PR into parquet-testing need to be merged before this is?

Not strictly necessary, but allows for better testing. We've expanded the decryption tests to read two sets of files - one produced in real time by the encryption tests; another stored in  the parquet-testing repo. The former can miss problems caused by matching bugs in writers and readers.

File: cpp/src/parquet/column_reader.cc
Comment (ggershinsky) [inline cpp/src/parquet/column_reader.cc]: Hi Tham, I believe we need to try to get rid of all #ifdef PARQUET_ENCRYPTION statements (if possible). Lets go over the code early next week (our weekend starts today..).

File: cpp/src/parquet/file_reader.cc
Comment (ggershinsky) [reply cpp/src/parquet/file_reader.cc]: missed this question, please see the comment below.

**Comment by wesm** [comment]
OK. I'll wait for you to address the ifdefs

**Comment by thamht4190** [comment]
@wesm @majetideepak we've figured out a way to remove all #ifdef PARQUET_ENCRYPTION. To be able to compile without openssl, we create a encryption_internal-nossl.cc instead of encryption_internal.cc with openssl. I'll update the fix, after internal team review, I will push to this pull.

**Comment by wesm** [comment]
Note that "internal review" is not really consistent with the Apache Way

**Comment by thamht4190** [comment]
Ok @wesm. So I will push the change directly to this pull.

**Comment by ggershinsky** [comment]
@wesm @majetideepak We are done with all todo items on our current list. This PR should be able to pass the CI now (at least the encryption unitests), due to availability of the new parquet-testing files (thanks Wes for merging them). Is there a way to trigger the CI here directly? Or we should add a new commit / rebase the PR?

**Comment by ggershinsky** [comment]
In parallel, @thamht4190 and @revit13 , if you're online now, can you check if this PR can be rebased (a good chance the Arrow master has moved..)

**Comment by thamht4190** [comment]
@majetideepak CI tests have passed except for R in AppVeyor that needs OpenSSL. It seems we should link openssl libs at line: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/26093493/job/9yen567yr3pw6fh8#L3403. However, I'm not sure how to do it. Can you please take a look?

**Comment by thamht4190** [comment]
We got this issue when testing our product on Windows with MacAfee (it may happen with MacOS as well):
`"Failed to open local file: <path_to_folder>/<file_name>.parquet.encrypted , error: Permission denied"`
Then when checking log of MacAfee, I see:
`NT AUTHORITY\SYSTEM ran CORTEXSERVICE.EXE, which tried to access <PATH_TO_FOLDER>\<FILE_NAME>.PARQUET.ENCRYPTED, violating the rule "*.encrypted Blocker", and was blocked. For information about how to respond to this event, see KB85494.`
It seems .encrypted extension is not so friendly with antivirus softwares.
As I know, parquet tools should not base on file extension, then it's quite safe to use another file extension (without .encrypted) in my application when writing parquet file, however, it seems tricky. Otherwise, it's inconvenient for our customers if keeping .encrypted extension. We are discussing to trade off. Any thoughts? @majetideepak

**Comment by ggershinsky** [comment]
adding @wesm to the loop on the last comment. My personal take on that - technically, you can use files without .encrypted extension in your settings; neither encryption spec nor the code depend on file names (and there are no plans today to introduce such dependency), so parquet tools should be able to read these files in the future.
However, I believe the right thing to do is to re-configure your anti-virus system (to enable say "*.parquet.encrypted" files). These files are encrypted after all.., so that should be handled at a policy level. But, eventually, this is up to the organizational security admin decision; naming these files without ".encrypted" extension can be also an option.

**Comment by ggershinsky** [comment]
@wesm @majetideepak We are done, let us know what's next. All comments have been addressed; the CI tests pass (except for one, R/libarrow, that lacks OpenSSL -  Deepak, your help with that will be appreciated).

**Comment by majetideepak** [comment]
@ggershinsky I will make a final pass today and tomorrow. I will also look into the R issue.

**Comment by majetideepak** [comment]
To fix the R failure, we have to link the OpenSSL library here https://github.com/apache/arrow/blob/48ee38f833a51bb46a404da10d89c4b687e952c9/r/src/Makevars.win
or disable Parquet Encryption in this test.
An issue in the current form is that there is no way to disable Parquet Encryption if the system has OpenSSL. I think we have to switch the default PARQUET_REQUIRE_ENCRYPTION to ON and support OFF. I will open a separate JIRA and make that change.

**Comment by majetideepak** [comment]
I decided to leave the current behavior of building encryption support if the system has OpenSSL libraries for now. This was discussed in https://github.com/apache/arrow/pull/4494
We can always change this behavior in the future.

**Comment by majetideepak** [comment]
@thamht4190 can you rebase with master again. The failing rust tests have been fixed.

File: ci/windows-pkg-arrow-for-r.sh
Comment (majetideepak) [inline ci/windows-pkg-arrow-for-r.sh]: CC: @nealrichardson

Comment (nealrichardson) [reply ci/windows-pkg-arrow-for-r.sh]: Thanks, that seems ok (though you need to remove the extra `:` in the middle of the URLs).

Since you had to make this change, I suspect that this means we'll need to update the Homebrew formula too, but unfortunately we don't have that under CI.

Comment (majetideepak) [reply ci/windows-pkg-arrow-for-r.sh]: Fixed the URLs. What are your thoughts on updating the Homebrew formula?

Comment (nealrichardson) [reply ci/windows-pkg-arrow-for-r.sh]: IMO we should put the C++ homebrew formula in our `ci/` directory and have a Travis job, perhaps only run nightly/on-demand if we're concerned about build time, that builds from it and then runs some test suite that requires it (e.g. R, Ruby). https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md suggests we can `brew install --build-from-source` so IIUC we'd use something like that. 

Related, I previously tried to add `brew install --HEAD` to CI but that didn't go anywhere: https://github.com/apache/arrow/pull/4576

I don't think this needs to block this patch, but it would be great to do something with this sooner rather than later. Otherwise we'll keep making releases unnecessarily expensive and that will keep us from doing them as frequently as we should.

**Comment by thamht4190** [comment]
The CI is green now. Thanks @majetideepak for handling the OpenSSL.
@majetideepak  @wesm The pull request is complete and ready for merge, can you please have a look?

**Comment by majetideepak** [comment]
@thamht4190 I will complete my review by tomorrow.

**Comment by majetideepak** [comment]
@thamht4190 I am running late on this. I will try to wrap up this week.

**Comment by thamht4190** [comment]
> 
> 
> @thamht4190 I am running late on this. I will try to wrap up this week.
Thanks!

File: cpp/src/parquet/encryption.h
Comment (philhoch) [inline cpp/src/parquet/encryption.h]: Is `explicit Builder(const std::string& name) : Builder(schema::ColumnPath::FromDotString(name), true) {}` intended instead?

File: cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc
Comment (majetideepak) [inline cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc]: Why not have another `parquet::FileEncryptionProperties::Builder` constructor that does not take a `footer_key` and infer `plaintext_footer` from the absence of the footer key?
From a user API point of view, the current requirement of needing a footer-key even if the user wants plaintext footer seems odd.

File: cpp/src/parquet/encryption.cc
Comment (majetideepak) [inline cpp/src/parquet/encryption.cc]: simply `disable_aad_prefix_storage` ?

File: cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc
Comment (majetideepak) [inline cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc]: should we have one option without any `FileDecryptionProperties` to test legacy reading?

File: cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc
Comment (majetideepak) [inline cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc]: I tried reading this file with the current master and got the following error. Can you take a look?
```
--- Values ---
boolean_field                 |int32_field                   |float_field                   |double_field                  |
1                             |0                             |Parquet error: Couldn't deserialize thrift: don't know what type: 
Deserializing page header failed.

```

Comment (majetideepak) [reply cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc]: This branch version says
```
./parquet-reader ./encrypt/tester3.parquet.encrypted 
........ File Metadata ......
Parquet error: Cannot decrypt ColumnMetadata. FileDecryptionProperties must be provided.
```

File: cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc
Comment (ggershinsky) [reply cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc]: Per the spec, footer key is always required - in plaintext footer mode, it is used to sign / tamper-proof the footer. We'll add an API comment to make this clear.

File: cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc
Comment (ggershinsky) [reply cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc]: good point, we should. we'll add this.

File: cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc
Comment (ggershinsky) [reply cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc]: the first error is exactly what should happen when a legacy reader tries to access an encrypted column (it gets a thrift exception on page header).
the second error is from a new reader, trying to read encrypted file without decryption properties.

File: cpp/src/parquet/encryption.cc
Comment (thamht4190) [reply cpp/src/parquet/encryption.cc]: yes, that's simpler. I'll fix it.

File: cpp/src/parquet/test-encryption-util.cc
Comment (majetideepak) [inline cpp/src/parquet/test-encryption-util.cc]: why not put this in the header itself?

File: cpp/src/parquet/encryption.h
Comment (majetideepak) [inline cpp/src/parquet/encryption.h]: `ColumnDecryptionProperties` already has the `schema::ColumnPath`. What is the purpose of this map?

File: cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc
Comment (majetideepak) [reply cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc]: Any reason to enforce decryption properties?
Why cannot we read encrypted parquet files with plaintext footer in legacy mode with the new reader?

File: cpp/src/parquet/encryption.h
Comment (majetideepak) [reply cpp/src/parquet/encryption.h]: If the purpose it to have a fast look up, we can probably keep a sorted vector of `schema::ColumnPath` and use [binary search](http://www.cplusplus.com/reference/algorithm/binary_search/)

File: cpp/src/parquet/encryption.h
Comment (majetideepak) [inline cpp/src/parquet/encryption.h]: It would have been much cleaner if we tied these to the `ColumnProperties` class. Just like `encoding` and `compression`, `encryption` is just another property for each column.
I can help with this if you need. It is much easier to clean up the interfaces before we merge.

File: cpp/src/parquet/encryption.h
Comment (ggershinsky) [reply cpp/src/parquet/encryption.h]: yep, the idea is to create these properties once for each relevant column, and to re-use them later. This doesn't happen much though, maybe twice or so. Also, typically the number of encrypted columns in not high. In any case, this is not in the critical performance path - because the main bottlenecks are data processing, and an interaction with a remote KMS . So we thought a map would be good enough.

File: cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.h
Comment (ggershinsky) [reply cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.h]: sounds good, we'll do that.

File: cpp/src/parquet/test-encryption-util.cc
Comment (ggershinsky) [reply cpp/src/parquet/test-encryption-util.cc]: makes sense to me.

File: cpp/src/parquet/encryption.h
Comment (ggershinsky) [reply cpp/src/parquet/encryption.h]: They are similar, but there are some differences. Eg encoding and compression are typically a single value, while encryption props are a structure with keys, key metadata, etc. Also, decryption properties are significantly different from encryption properties. Plus, we'd like to keep encryption API similar / identical in C++ and Java versions - https://github.com/apache/parquet-mr/pull/614

Comment (majetideepak) [reply cpp/src/parquet/encryption.h]: Can we rename all variables with name `column_properties` in this patch to `column_encryption_properties` to clearly distinguish them?

File: cpp/src/parquet/encryption.h
Comment (majetideepak) [reply cpp/src/parquet/encryption.h]: The map is fine. Looking at the usage, should the key be "std::string columnPath" instead? `CmpColumnPath` is converting them to strings anyway

File: cpp/src/parquet/encryption.h
Comment (ggershinsky) [reply cpp/src/parquet/encryption.h]: sure, we'll do that.

File: cpp/src/parquet/encryption.h
Comment (ggershinsky) [reply cpp/src/parquet/encryption.h]: ok, we can do that.

File: cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc
Comment (ggershinsky) [reply cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc]: per the previous comment, we are adding a test option without (with null) FileDecryptionProperties to read encrypted parquet files with plaintext footer in legacy mode with the new reader

File: cpp/src/parquet/encryption.h
Comment (ggershinsky) [reply cpp/src/parquet/encryption.h]: Hi Deepak, we've started working on that, but turns out this has two sides. The API, and internal SPI, work with schema::ColumnPath objects, so this conversion happens one way or another. Fortunately, this is a low frequency operation, so no effect on performance. I hope you're ok with leaving this code as is.

**Comment by thamht4190** [comment]
Hi @majetideepak, we've addressed all comments. I'm trying to fix CI build issue, however, there is still one failure that I don't know how to fix: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/26860281/job/55qm48aefgrveq2q#L3418.
It seems because of OpenSSL missing. Can you please help me fix it or point me out how to fix it? Thanks!

**Comment by wesm** [comment]
cc @nealrichardson @kou since this is happening on the MinGW / R side

**Comment by wesm** [comment]
also @jeroen

File: r/configure.win
Comment (nealrichardson) [inline r/configure.win]: This looks like a bad merge with master; r/configure.win has changed since this branch was started. L46-49 below overwrites these.

File: ci/windows-pkg-arrow-for-r.sh
Comment (nealrichardson) [inline ci/windows-pkg-arrow-for-r.sh]: This may not work, you may need a "backports" version of openssl. @jeroen might know.

Comment (nealrichardson) [reply ci/windows-pkg-arrow-for-r.sh]: Also I'd be surprised that you could add a dependency and not have to edit the depends statement in `ci/PKGBUILD`: https://github.com/apache/arrow/blob/master/ci/PKGBUILD#L27

**Review by nealrichardson — CHANGES_REQUESTED** [review]
I added a couple of notes. What if you just disabled encryption for the R Windows build for now, if this is still a sticking point? The libarrow that ships with the R package on Windows is slimmer than most. Presumably there's a cmake flag for turning that off?

File: ci/windows-pkg-arrow-for-r.sh
Comment (thamht4190) [reply ci/windows-pkg-arrow-for-r.sh]: Any idea @majetideepak?

Comment (thamht4190) [reply ci/windows-pkg-arrow-for-r.sh]: > Presumably there's a cmake flag for turning that off?
Not yet. Encryption is turn on as default when openssl is found. Any idea @majetideepak?

File: r/configure.win
Comment (thamht4190) [reply r/configure.win]: Right. I'll fix it.

File: ci/windows-pkg-arrow-for-r.sh
Comment (majetideepak) [reply ci/windows-pkg-arrow-for-r.sh]: Looks like we should add a flag to disable encryption support even if the system has OpenSSL.
@thamht4190 Can you rebase with master first?

Comment (thamht4190) [reply ci/windows-pkg-arrow-for-r.sh]: @majetideepak the rebase would take a lot of effort so I've merged latest code to my code.

**Comment by wesm** [comment]
@thamht4190 the `Encryptor` class is missing `PARQUET_EXPORT`

**Comment by thamht4190** [comment]
wow. Thanks @wesm. All builds are green now. cc @majetideepak

File: ci/windows-pkg-arrow-for-r.sh
Comment (majetideepak) [reply ci/windows-pkg-arrow-for-r.sh]: @nealrichardson there is a green build now. Do you still prefer to disable OpenSSL for Windows and R? The current state is that OpenSSL get included by default if CMake finds it in the environment (and Parquet build is ON).

Comment (nealrichardson) [reply ci/windows-pkg-arrow-for-r.sh]: If the build is passing, I'm ok with leaving it on by default. Is there a way to turn it off if we do encounter unforeseen circumstances when submitting to CRAN?

**Comment by wesm** [comment]
Would be great to get this merged soon. Looks like another rebase is going to be required

File: ci/windows-pkg-arrow-for-r.sh
Comment (majetideepak) [reply ci/windows-pkg-arrow-for-r.sh]: We don't and I also agree that we should have one. I will make this change.

**Comment by thamht4190** [comment]
I'll wait for @majetideepak's review, then update code along with rebase.

File: ci/windows-pkg-arrow-for-r.sh
Comment (majetideepak) [reply ci/windows-pkg-arrow-for-r.sh]: encryption wont be built by default unless `PARQUET_REQUIRE_ENCRYPTION` is specified.

File: cpp/src/parquet/column_reader.cc
Comment (majetideepak) [inline cpp/src/parquet/column_reader.cc]: I refactored this a bit. Let me know if you have any comments.

File: cpp/src/parquet/encryption.h
Comment (majetideepak) [reply cpp/src/parquet/encryption.h]: I will be at Strata NY next two days, but after I come back I want to take a look at this.

**Comment by wesm** [comment]
What is the status of this? It seems the testing submodule needs to be updated

**Comment by majetideepak** [comment]
@wesm I plan to work on one last round of refactoring this week and we should merge it early next week. We can continue to make further code improvements after we merge. I will resolve the conflict.

**Comment by majetideepak** [comment]
@wesm I am done with my refactoring for now. We can merge this PR if there are no other comments.

**Comment by majetideepak** [comment]
All tests are passing. @thamht4190, @revit1976 @ggershinsky do you have any comments on this PR? Is this good to be merged?

**Comment by ggershinsky** [comment]
no additional comments, looks good to me.

**Comment by thamht4190** [comment]
LGTM. Thanks Deepak for your change.

**Comment by revit1976** [comment]
LGTM also. Thanks.

**Comment by majetideepak** [comment]
@wesm Do you plan to make a pass before we merge this?

**Comment by wesm** [comment]
@majetideepak sorry for the delay. I am going to take a brief look through this to check for anything that seems obviously off

**Comment by wesm** [comment]
I apologize, I've been out of pocket the last couple weeks. Taking a look through and if I don't any anything troubling I'll merge

File: cpp/src/parquet/encryption_read_configurations_test.cc
Comment (wesm) [inline cpp/src/parquet/encryption_read_configurations_test.cc]: A couple non-conforming names in this file

File: cpp/src/parquet/file_reader.cc
Comment (wesm) [inline cpp/src/parquet/file_reader.cc]: I suspect that some code in these methods could be simplified or reused

File: dev/release/verify-release-candidate.sh
Comment (wesm) [inline dev/release/verify-release-candidate.sh]: This may make it harder for people to verify releases. We can see for the next major release

**Review by wesm — APPROVED** [review]
+1. I'm going to push some stylistic fixes. It's a bummer that this branch is so messed up. I'm going to create a squashed version to merge to master that doesn't have all the unrelated commits in it. I'll have to manually list out the co-authors...

**Comment by wesm** [comment]
Here's the merge-ready commit, I'm going to wait a bit for CI to make progress

https://github.com/wesm/arrow/commit/69db453808cb8afacb9bc541bbd14d13a7dd1372

Do NOT merge this PR as is, please

**Comment by ggershinsky** [comment]
@majetideepak @wesm While Parquet encryption feature is a part of the Arrow 0.16.0 release, it doesn't appear in https://github.com/apache/arrow/blob/master/CHANGELOG.md. Also, the https://issues.apache.org/jira/browse/PARQUET-1300 is still open. Should it be closed, with setting the "Affects Version/s:" field? Can the 0.16.0 changelog be updated?

**Comment by majetideepak** [comment]
@ggershinsky I think if needed we could do a second minor release ARROW 0.16.1 if there are no objections. Do you want to e-mail the Arrow/Parquet mailing lists to discuss this?

**Comment by ggershinsky** [comment]
@majetideepak We've submitted a paper where we've referenced Arrow 0.16.0 for parquet encryption. So having it in 0.16.1 doesn't solve our immediate (though very minor) problem. I wouldn't suggest running a new release just for adding this to the log. I believe we can think of 0.16.0 as a "silent release" of this feature.  I've created an Arrow jira (8018) for adding this feature explicitly in 1.0.0, and closed the PARQUET-1300 as fixed - but if I've done something wrong there, please change as needed.

**Comment by ggershinsky** [comment]
@wesm what should be done for packaging issues for encryption? I would handle this, but not 100% sure about the details.

**Comment by wesm** [comment]
I just opened ARROW-8038 and ARROW-8040 to cover some of the basic packaging stuff. Let's move the discussion there (and to child issues)

**Pull request closed without merging**
